src/tools/tidy/src/fluent_used.rs

```diff
@@ -0,0 +1,36 @@
+//! Checks that all Fluent messages appear at least twice
+
+use crate::walk::{filter_dirs, walk};
+use regex::Regex;
+use std::collections::HashMap;
+use std::path::Path;
+
+lazy_static::lazy_static! {
+    static ref WORD: Regex = Regex::new(r"\w+").unwrap();
+}
+
+fn filter_used_messages(
+    contents: &str,
+    msgs: &mut HashMap<String, String>,
+    unused_msgs: &mut HashMap<String, String>,
+) {
+    let mut matches = WORD.find_iter(contents);
+    while let Some(name) = matches.next() {
+        if let Some((name, filename)) = msgs.remove_entry(name.as_str()) {
+            unused_msgs.insert(name, filename);
+        } else {
+            unused_msgs.remove(name.as_str());
+        }
```

Comment:
This is quite complicated and counterintuitive and it took me a while until I realized the reason this is correct (it sees every message at least once, as it also looks at ftl files, so if it only appears once it's unused) (a comment would have helped^^).

Suggestion for an easier to understand algorithm: When walking the paths, only check Rust files and then have the simple algorithm of starting with a hash map with all messages and try to remove every word. The remaining messages are unused.

Reply:
But some messaged can be used as a part of other messages, like `parse_sugg_change_inner_attr_to_outer` in the following:
```
parse_sugg_change_inner_attr_to_outer = to annotate the {$item}, change the attribute from inner to outer style

parse_inner_attr_not_permitted = an inner attribute is not permitted in this context
    .label_does_not_annotate_this = {parse_label_inner_attr_does_not_annotate_this}
    .sugg_change_inner_to_outer = {parse_sugg_change_inner_attr_to_outer}

parse_inner_attr_not_permitted_after_outer_attr = an inner attribute is not permitted following an outer attribute
    .label_attr = not permitted following an outer attribute
    .label_prev_attr = previous outer attribute
    .label_does_not_annotate_this = {parse_label_inner_attr_does_not_annotate_this}
    .sugg_change_inner_to_outer = {parse_sugg_change_inner_attr_to_outer}

parse_inner_attr_not_permitted_after_outer_doc_comment = an inner attribute is not permitted following an outer doc comment
    .label_attr = not permitted following an outer doc comment
    .label_prev_doc_comment = previous doc comment
    .label_does_not_annotate_this = {parse_label_inner_attr_does_not_annotate_this}
    .sugg_change_inner_to_outer = {parse_sugg_change_inner_attr_to_outer}
```

Reply:
It is unused in Rust files but in Fluent files.

Reply:
Ah... in that case a comment may have to be enough